[quant] Enable reduce_range for graphmode

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39874 [quant] Enable reduce_range for graphmode**

Summary:
When fbgemm backend is set we make sure reduce_range is set to true to avoid overflow in the operator
Also adds test for per-channel quant with graph mode and compare numerics with eager mode

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22011205](https://our.internmc.facebook.com/intern/diff/D22011205)

supriyar committed:
[quant] Enable reduce_range for graphmode
Summary:
When fbgemm backend is set we make sure reduce_range is set to true to avoid overflow in the operator
Also adds test for per-channel quant with graph mode and compare numerics with eager mode

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -70,7 +70,11 @@ bool isWeight(Module& module, Value* v) {
   std::string choose_qparams_func = "_choose_qparams_per_tensor";
-  auto reduce_range = graph->insertConstant(false);
+  bool reduce_range_param = false;
+  if (at::globalContext().qEngine() == at::QEngine::FBGEMM) {
```
this means we have an implicit argument of qengine, which makes the produced quantized model tied to fbgemm, which is not desirable, is there any plan to remove this in the future?

supriyar replied:
This argument should not affect the final converted quantized model. The `reduce_range` option gets added to `linear_dynamic` op as an argument. During runtime if backend is not fbgemm, it gets ignored.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L20
and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L222

jerryzh168 replied:
what about the other way around? if we produce the model with qengine set to qnnpack, does it run on fbgemm qengine?

supriyar replied:
It will run, but it might not be the most optimal. In the future we can make reduce_range default to true for the op, but that might break BC for numerics of existing models.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -70,7 +70,11 @@ bool isWeight(Module& module, Value* v) {
+  if (at::globalContext().qEngine() == at::QEngine::FBGEMM) {
+    reduce_range_param = true;
+  }
+  auto reduce_range = graph->insertConstant(reduce_range_param);
```
you probably forgot to insert this as an argument for some node?

supriyar replied:
This is already being used in the choose_qparams node below.

jerryzh168 replied:
oh I didn't see the deleted line on the left

supriyar committed:
Update on "[quant] Enable reduce_range for graphmode"


Summary:
When fbgemm backend is set we make sure reduce_range is set to true to avoid overflow in the operator
Also adds test for per-channel quant with graph mode and compare numerics with eager mode

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Enable reduce_range for graphmode"


Summary:
When fbgemm backend is set we make sure reduce_range is set to true to avoid overflow in the operator
Also adds test for per-channel quant with graph mode and compare numerics with eager mode

Test Plan:
python test/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22011205](https://our.internmc.facebook.com/intern/diff/D22011205)

[ghstack-poisoned]
